pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,316 @@
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionNoGardenFound, "no garden found")
+		if err := r.updateExtensionStatus(ctx, log, extension, conditions); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error updating extension status: %w", err)
+		}
+		return reconcile.Result{}, r.removeFinalizers(ctx, log, extension)
+	}
+
+	garden := &gardenList.Items[0]
+	gardenClientSet, err := r.GardenClientMap.GetClient(ctx, keys.ForGarden(garden))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving garden client object: %w", err)
+	}
+
+	if extension.DeletionTimestamp != nil {
+		return reconcile.Result{}, r.delete(ctx, log, gardenClientSet.Client(), extension)
+	}
+	return reconcile.Result{}, r.reconcile(ctx, log, gardenClientSet.Client(), extension)
+}
+
+func (r *Reconciler) updateExtensionStatus(ctx context.Context, log logr.Logger, extension *operatorv1alpha1.Extension, updatedConditions VirtualClusterConditions) error {
+	currentConditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+	if extension.Generation == extension.Status.ObservedGeneration && !v1beta1helper.ConditionsNeedUpdate(currentConditions.ConvertToSlice(), updatedConditions.ConvertToSlice()) {
+		return nil
+	}
+
+	patch := client.MergeFrom(extension.DeepCopy())
+	// Rebuild garden currentConditions to ensure that only the currentConditions with the correct types will be updated, and any other
+	// currentConditions will remain intact
+	extension.Status.Conditions = v1beta1helper.BuildConditions(extension.Status.Conditions, updatedConditions.ConvertToSlice(), currentConditions.ConditionTypes())
+	extension.Status.ObservedGeneration = extension.Generation
+
+	// prevent sending empty patches
+	if data, err := patch.Data(extension); err != nil {
+		return fmt.Errorf("failed getting patch data for Extension %s: %w", extension.Name, err)
+	} else if string(data) == "{}" {
+		return nil
+	}
+
+	log.V(1).Info("Updating extension status currentConditions")
+	if err := r.RuntimeClient.Status().Patch(ctx, extension, patch); err != nil {
+		log.Error(err, "Could not update extension status")
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	conditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+	if !controllerutil.ContainsFinalizer(extension, operatorv1alpha1.FinalizerName) {
+		log.Info("Adding finalizer")
+		if err := controllerutils.AddFinalizers(ctx, r.RuntimeClient, extension, operatorv1alpha1.FinalizerName); err != nil {
+			return fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	log.Info("Reconciling extension virtual resources", "name", extension.Name)
+	err := r.reconcileVirtualClusterResources(ctx, log, gardenClient, extension)
+	if err != nil {
```

Comment:
combine

Reply:
I don't think I can here, because I join the error couple lines later with a (potential) error from the update status call.